File: DSLib/Lib/SegmentTree/SegmentTree.cpp
#include "SegmentTree.h"

namespace ds
{
	namespace bin_tree
	{
		seg_tree::seg_tree(const void* container_cookie, const std::size_t size, node<std::size_t>* (*merge_nodes)(const void*, node<std::size_t>*, node<std::size_t>*)) : container_cookie(container_cookie), merge_nodes(merge_nodes), size(size)
		{
			const range segment{ 0, size - 1 };
			this->root = build_tree(node_type::k_root, segment);
		}

		node<std::size_t>* seg_tree::query(const range& query_segment) const
		{
			const range segment{ 0, size - 1 };
			return query(this->root, segment, query_segment);
		}

		node<std::size_t>* seg_tree::build_tree(const node_type type, const range& segment) const
		{
			if (segment.lower_bound == segment.upper_bound)
			{
				return new node<std::size_t>(segment.lower_bound, type);
			}

			range new_segment;
			new_segment.lower_bound = segment.lower_bound;
			new_segment.upper_bound = (segment.lower_bound + segment.upper_bound) >> 1;
			const auto left_child = build_tree(node_type::k_left_child, new_segment);

			new_segment.lower_bound = new_segment.upper_bound + 1;
			new_segment.upper_bound = segment.upper_bound;
			const auto right_child = build_tree(node_type::k_right_child, new_segment);

			const auto new_node = new node<std::size_t>(merge_nodes(container_cookie, left_child, right_child)->data, type);
			new_node->left_child = left_child;
			new_node->right_child = right_child;
			left_child->ancestor = right_child->ancestor = new_node;
			return new_node;
		}


		node<std::size_t>* seg_tree::query(node<std::size_t>* n, const range& segment, const range& query_segment) const
		{
			if (query_segment.lower_bound > segment.upper_bound || query_segment.upper_bound < segment.lower_bound)
			{
				return nullptr;
			}

			if (segment.lower_bound >= query_segment.lower_bound && segment.upper_bound <= query_segment.upper_bound)
			{
				return n;
			}

			range new_segment;
			new_segment.lower_bound = segment.lower_bound;
			new_segment.upper_bound = (segment.lower_bound + segment.upper_bound) >> 1;
			const auto left_child = query(n->left_child, new_segment, query_segment);

			new_segment.lower_bound = new_segment.upper_bound + 1;
			new_segment.upper_bound = segment.upper_bound;
			const auto right_child = query(n->right_child, new_segment, query_segment);

			if (left_child == nullptr)
			{
				return right_child;
			}

			if (right_child == nullptr)
			{
				return left_child;
			}

			return merge_nodes(container_cookie, left_child, right_child);
		}
	}
}

File: DSLib/Lib/SegmentTree/SegmentTree.java
package Lib.SegmentTree;

import Lib.BinaryTree.BinaryTree;
import Lib.Tree.Node;
import Lib.Tree.NodeType;

import java.util.HashMap;

public class SegmentTree<T> extends BinaryTree<T> {
    public SegmentTree(Object collectionCookie, int size, AccessCollection<T> accessCollection, MergeNodes<T> mergeNodes) {
        this.collectionCookie = collectionCookie;
        this.size = size;
        this.accessCollection = accessCollection;
        this.mergeNodes = mergeNodes;
        lazyStore = new HashMap<>();

        Range segment = new Range(0, size - 1);
        super.root = buildTree(NodeType.ROOT, segment);
    }

    public Response<T> query(Range querySegment) {
        Range segment = new Range(0, size - 1);
        return query(super.root, segment, querySegment);
    }

    private Response<T> query(Node<T> n, Range segment, Range querySegment) {
        if ((querySegment.lowerBound > segment.upperBound) || (querySegment.upperBound < segment.lowerBound)) {
            return new Response<>(false, null);
        }

        if (lazyStore.containsKey(n)) {
            T lazyData = lazyStore.get(n);
            n.data = mergeNodes.merge(n.data, lazyData);
            propagateLaziness(n, lazyData);
            lazyStore.remove(n);
        }

        if ((segment.lowerBound >= querySegment.lowerBound) && (segment.upperBound <= querySegment.upperBound)) {
            return new Response<>(true, n.data);
        }

        Range leftSegment = new Range(segment.lowerBound, (segment.lowerBound + segment.upperBound) >> 1);
        Response<T> leftResponse = query(n.leftChild, leftSegment, querySegment);

        Range rightSegment = new Range(leftSegment.upperBound + 1, segment.upperBound);
        Response<T> rightResponse = query(n.rightChild, rightSegment, querySegment);

        if (!leftResponse.isValid) {
            return rightResponse;
        }

        if (!rightResponse.isValid) {
            return leftResponse;
        }

        return new Response<>(true, mergeNodes.merge(leftResponse.data, rightResponse.data));
    }

    private void propagateLaziness(Node<T> n, T data) {
        if (n.leftChild != null && n.rightChild != null) {
            return;
        }

        if (lazyStore.containsKey(n.leftChild)) {
            T lazyData = lazyStore.get(n.leftChild);
            lazyStore.put(n.leftChild, mergeNodes.merge(lazyData, data));
        } else {
            lazyStore.put(n.leftChild, data);
        }

        if (lazyStore.containsKey(n.rightChild)) {
            T lazyData = lazyStore.get(n.rightChild);
            lazyStore.put(n.rightChild, mergeNodes.merge(lazyData, data));
        } else {
            lazyStore.put(n.rightChild, data);
        }
    }

    private Node<T> buildTree(NodeType type, Range segment) {
        if (segment.lowerBound == segment.upperBound) {
            return new Node<>(accessCollection.access(collectionCookie, segment.lowerBound), type);
        }

        Range leftSegment = new Range(segment.lowerBound, (segment.lowerBound + segment.upperBound) >> 1);
        Node<T> leftChild = buildTree(NodeType.LEFT_CHILD, leftSegment);

        Range rightSegment = new Range(leftSegment.upperBound + 1, segment.upperBound);
        Node<T> rightChild = buildTree(NodeType.RIGHT_CHILD, rightSegment);

        Node<T> parent = new Node<>(mergeNodes.merge(leftChild.data, rightChild.data), type);
        parent.leftChild = leftChild;
        parent.rightChild = rightChild;
        leftChild.ancestor = rightChild.ancestor = parent;
        return parent;
    }

    public Object collectionCookie;
    public int size;
    public MergeNodes<T> mergeNodes;
    public AccessCollection<T> accessCollection;

    private HashMap<Node<T>, T> lazyStore;
}
